src/postprocessor/chunk.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2022, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Chunking postprocessor that concatenates chunks of data, until its `max_bytes` is reached,
+//! then it releases all the data as one chunk (`Vec<u8>`).
+
+use super::Postprocessor;
+use crate::errors::{Error, Kind as ErrorKind, Result};
+use tremor_script::Value;
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+struct Config {
+    max_bytes: usize,
+}
+
+pub(crate) struct Chunk {
+    max_bytes: usize,
+    chunk: Vec<u8>,
+}
+
+impl Postprocessor for Chunk {
+    fn name(&self) -> &str {
+        "chunk"
+    }
+
+    fn process(&mut self, _ingres_ns: u64, _egress_ns: u64, data: &[u8]) -> Result<Vec<Vec<u8>>> {
+        if (self.chunk.len() + data.len()) > self.max_bytes {
+            if data.len() > self.max_bytes {
+                self.warn(data.len());
+                Ok(vec![])
+            } else {
+                let mut output = Vec::with_capacity(self.max_bytes);
+                // append the new data to the buffer we do keep
+                output.extend_from_slice(data);
+                std::mem::swap(&mut output, &mut self.chunk);
+                Ok(vec![output])
+            }
+        } else {
+            self.chunk.extend_from_slice(data);
+            Ok(vec![])
+        }
```

Comment:
This would make the error condition a bit clearer and de-nest the expression
```suggestion
        if data.len() > self.max_bytes {
            self.warn(data.len());
            Ok(vec![])
        } else if (self.chunk.len() + data.len()) > self.max_bytes {
            let mut output = Vec::with_capacity(self.max_bytes);
            // append the new data to the buffer we do keep
            output.extend_from_slice(data);
            std::mem::swap(&mut output, &mut self.chunk);
            Ok(vec![output])
        } else {
            self.chunk.extend_from_slice(data);
            Ok(vec![])
        }
```